File: programs/Hamilt1.m
function [Beta_out, m_out, m, yobs_tmp]=Hamilt1(chol_index, Phi_temp,...
                                        j, yobs, tau_temp, Beta_temp, xi_temp)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Does HMC updates for the coefficents that are different
%
%   Input:
%       1) chol_index - s index matrix
%       2) Phi_temp - indicate variable that determine which coefficient
%       should be different
%       3) j - one of two segments
%       4) yobs - time series observations 
%       5) tau_temp - current smoothing parameters
%       6) Beta_temp - current coefficients
%       7) xi__temp - current partitions
%   Main Outputs:
%       1) Beta_out - vector of coefficients 
%       2) m_out - updated momentum variables
%       3) m - initial momentum variables
%       4) yobs_tmp - time series observations in selected segment
%
%   Required programs:  Gradient1
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%  
    global nbasis sigmasqalpha dimen nBeta M ee

    %pick right portion of the data
    if j>1
        yobs_tmp = yobs((xi_temp(j-1)+1):xi_temp(j),:);
    else
        yobs_tmp = yobs(1:xi_temp(j),:);
    end
    
    select = chol_index(Phi_temp,:).*(1:dimen^2);
    select = select(select~=0);
    
    
    ll = nBeta*length(select);
    [gr]=Gradient1(yobs_tmp, chol_index, Phi_temp, tau_temp,...
                                    Beta_temp, sigmasqalpha, nbasis);
    Beta_old = reshape(Beta_temp(:,select),ll,1); 
    Beta_out = Beta_old;
    
    % generate momentum variable
    m = mvnrnd(zeros(ll,1),M*eye(ll))';
    % determine leap number and step
    stepsize = unifrnd(0,2*ee);
    leap = randsample(1:2*ceil(1/ee),1);
    %leap = randsample(1:(1/stepsize),1);
    
    m_out = m + 0.5*gr*stepsize;
    for i=1:leap
        Beta_out = Beta_out + stepsize*(1/M)*eye(ll)*m_out;
        Beta_temp(:,select) = reshape(Beta_out,nBeta,length(select));
        if i==leap
            [gr] = Gradient1(yobs_tmp, chol_index, Phi_temp, tau_temp,...
                                    Beta_temp, sigmasqalpha, nbasis);
            m_out = m_out + 0.5*gr*stepsize;
        else
            [gr] = Gradient1(yobs_tmp, chol_index, Phi_temp, tau_temp,...
                                    Beta_temp, sigmasqalpha, nbasis);
            m_out = m_out + 1*gr*stepsize;
        end    
    end    
    m_out = -m_out;
  

File: R programs/Hamilt1.R
 Hamilt1 <- function(chol_index, Phi_temp, 
                     j, yobs, tau_temp, Beta_temp, xi_temp, global_info){
         nbasis <- global_info$nbasis; sigmasqalpha <- global_info$sigmasqalpha; dimen <- global_info$dimen; nBeta <- global_info$nBeta
         M <- global_info$M; ee <- global_info$ee
         
         # pick right portion of the data
         if (j>1){
                 yobs_tmp <- yobs[(xi_temp[j-1]+1):xi_temp[j],]
         }else{
                 yobs_tmp <- yobs[1:xi_temp[j],]
         }
         
         select <- which(chol_index[Phi_temp,]!=0)
         
         ll <- nBeta*length(select)
         gr <- Gradient1(yobs_tmp,chol_index, Phi_temp, tau_temp, 
                                Beta_temp,sigmasqalpha, nbasis, global_info)
         Beta_old <- matrix(Beta_temp[,select],ll,1)
         Beta_out <- Beta_old

         # generate momentum variable
         m <- rmvnorm(1,  rep(0, nrow(diag(ll))), M*diag(ll))
         # determine leap number and step
         stepsize <- runif(1, 0, 2*ee)
         leap <- sample(1:(2*ceiling(1/ee)), 1, replace = TRUE)
         
         m_out<- m + 0.5*gr*stepsize
         for(i in 1:leap){
                 Beta_out <- Beta_out + stepsize*(1/M)*diag(ll)%*%t(m_out)
                 Beta_temp[,select] <- matrix(Beta_out, nBeta, length(select))
                 if(i==leap){
                         gr <- Gradient1(yobs_tmp, chol_index, Phi_temp, tau_temp, 
                                         Beta_temp, sigmasqalpha, nbasis, global_info)
                         m_out <- m_out + 0.5*gr*stepsize
                 }else{
                         gr <- Gradient1(yobs_tmp, chol_index, Phi_temp, tau_temp, 
                                         Beta_temp,sigmasqalpha, nbasis, global_info)
                         m_out <- m_out + 1*gr*stepsize
                 }
         }
         m_out <- -m_out
         list(Beta_out=Beta_out, m_out=m_out, m=m, yobs_tmp=yobs_tmp)
 }
 
 
 
 
 